core/src/main/scala/org/apache/spark/sql/execution/ObjectHashMapAccessor.scala

```diff
@@ -1260,23 +1294,32 @@ case class ObjectHashMapAccessor(@transient session: SnappySession,
   private def genVarAssignCode(colVar: String, resultVar: ExprCode,
       dataType: DataType, doCopy: Boolean): String = dataType match {
     // if doCopy is true, then create a copy of some non-primitives that just
-    // hold a reference to UnsafeRow bytes (and can change under the hood)
-    case StringType if doCopy =>
+    // holds a reference to UnsafeRow bytes (and can change under the hood)
+    case StringType if doCopy && !multiMap =>
       s"$colVar = ${resultVar.value}.getBytes();"
-    case StringType =>
-      // try to copy just reference of the byte[] if possible
+    case StringType if !multiMap =>
+      // copy just reference of the object if underlying byte[] is immutable
       val stringVar = resultVar.value
       val platformClass = classOf[Platform].getName
       val bytes = ctx.freshName("stringBytes")
-      val obj = bytes + "Obj"
-      s"""Object $obj; byte[] $bytes;
+      s"""byte[] $bytes;
         if ($stringVar.getBaseOffset() == $platformClass.BYTE_ARRAY_OFFSET
-            && ($obj = $stringVar.getBaseObject()) instanceof byte[]
-            && ($bytes = (byte[])$obj).length == $stringVar.numBytes()) {
+            && ($bytes = (byte[])$stringVar.getBaseObject()).length == $stringVar.numBytes()) {
           $colVar = $bytes;
         } else {
           $colVar = $stringVar.getBytes();
         }"""
+    // multimap holds a reference to UTF8String itself
+    case StringType if doCopy =>
```

Comment:
Can there be a case when doCopy= false but multimap=true? If yes, would the default case ("$colVar = ${resultVar.value};") work for it?